runtime/url_util.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::uri_error;
+use deno_core::error::AnyError;
+use deno_core::ModuleSpecifier;
+use std::path::PathBuf;
+
+/// Attempts to convert a specifier to a file path. By default, uses the Url
+/// crate's `to_file_path()` method, but falls back to try and resolve unix-style
+/// paths on Windows.
+pub fn specifier_to_file_path(
+  specifier: &ModuleSpecifier,
+) -> Result<PathBuf, AnyError> {
+  let result = if !specifier.as_str().starts_with("file:///") {
+    Err(())
+  } else if cfg!(windows) {
+    match specifier.to_file_path() {
+      Ok(path) => Ok(path),
+      Err(()) => {
+        // This might be a unix-style path which is used in the tests even on Windows.
+        // Attempt to see if we can convert it to a `PathBuf`. This code should be removed
+        // once/if https://github.com/servo/rust-url/issues/730 is implemented.
+        if specifier.scheme() == "file"
+          && specifier.host().is_none()
+          && specifier.port().is_none()
+          && specifier.path_segments().is_some()
+        {
+          let path_str = specifier.path();
+          match String::from_utf8(
+            percent_encoding::percent_decode(path_str.as_bytes()).collect(),
+          ) {
+            Ok(path_str) => Ok(PathBuf::from(path_str)),
+            Err(_) => Err(()),
+          }
+        } else {
+          Err(())
+        }
+      }
+    }
+  } else {
+    specifier.to_file_path()
+  };
+  match result {
+    Ok(path) => Ok(path),
+    Err(()) => Err(uri_error(format!(
+      "Invalid file path.\n  Specifier: {specifier}"
+    ))),
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use super::*;
+  #[test]
+  fn test_specifier_to_file_path() {
```

Comment:
Seems this test is failing on Mac.

Reply:
that's weird how is it bypassing ` let result = if !specifier.as_str().starts_with("file:///") {`

I'll have to debug on the ci

Reply:
so the gist is each platform have its own way to do this

- file:// errors on linux
- file:/ errors on windows
- mac os doesn't error on either

Reply:
This is what I ended up with https://github.com/denoland/deno/pull/20062/commits/0153af2af424bc0375132025ff74ab601a903f04

Reply:
I don't like that the method have different result based on the os, I probably need to improve this somehow

Reply:
Or maybe it can be considered an upstream issue ? since these differences are coming from `url::Url::parse`, or are urls platform dependent by nature?

Reply:
I skimmed the issues on url an it seems that its working as intended, the scheme is expected to have 3 slashes after it, I guess if I can make the mac os also error on this it would match the behavior of the others , or leave it be since that's how urls on those systems work